DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md

```diff
@@ -0,0 +1,206 @@
+
+<img src="merge_unit.png" alt="merge_unit" width="600"/>
+
+A single merge unit requires `lg(N)` iterations to sort `N` elements. This requires the host enqueuing `lg(N)` iterations of the merge unit kernels that merge sublists of size {`1`, `2`, `4`, ...} into larger lists of size {`2`, `4`, `8`, ...}, respectively. This results in a timeline that looks like the figure below.
+
+<img src="basic_runtime_graph.png" alt="basic_runtime_graph" width="800"/>
+
+To achieve SIMD level parallelism, we enhance the merge unit to merge `k` elements per cycle. The figure below shows how this is done. We will assume that we are sorting from largest to smallest, but the logic is very similar for sorting largest to smallest and is easily configurable at compile time in this design. <br/>
+
+The merge unit looks at the two inputs of size `k` coming from `ProducerA` and `ProducerB` (in the figure below, `k=4`) and compares the first elements of each set; remember, these set of `k` elements are already sorted, so we are comparing the smallest (or largest, depending on the sorting order) elements of the set. Whichever set of elements has the *smaller of the smallest elements* is chosen and combined with the `k` other elements from the feedback path. These `2*k` elements go through a merge sort network that sorts them. After the `2*k` elements are sorted, the smallest `k` elements are output (to the `Consume` kernel) and the largest `k` elements are fed back into the sorting network, and the process repeats. This allows the merge unit to process `k` elements per cycle in the steady state. note: `k` must be a power of 2. <br/>
+
+More information on this design can be found in this paper by [R. Kobayashi and K. Kise](https://www.researchgate.net/publication/316604001_A_High_Performance_FPGA-Based_Sorting_Accelerator_with_a_Data_Compression_Mechanism).
+
+<img src="k-way_merge_unit.png" alt="way_merge_unit" width="900"/>
+
+To achieve thread level parallelism, the merge sort design accepts a template parameter, `units`, which allows one to instantiate multiple instances of the merge unit, as shown in the figure below. Before the merge units start, the incoming data coming from the input pipe is sent through a bitonic sorting network and written to the temporary buffer partitions in device memory. This sorting network sorts `k` elements per cycle in the steady state. Choosing the number of merge units is an area-performance tradeoff (note: the number of instantiated merge units must be a power of 2). In this design, each merge unit sorts a partition of the input data of size `N/units`. 
+
+<img src="parallel_tree_bitonic_k-way.png" alt="parallel_tree_bitonic_k-way" width="800"/>
```

Comment:
Why do the merge units each have an arrow showing that they write to DDR? I don't see where this is mentioned in the Readme, did I miss it?

Reply:
Because the write to temp read buffers in device memory.
"The lists are streamed in from device memory (e.g., DDR or HBM) by two `Produce` kernels. The `Consumer` kernel can stream data to either a SYCL pipe or to device memory."